crates/kitsune2/src/factories/mem_peer_store.rs

```diff
@@ -0,0 +1,261 @@
+            .filter_map(|info| {
+                if info.is_tombstone {
+                    return None;
+                }
+
+                if !arcs_overlap(arc, info.storage_arc) {
+                    return None;
+                }
+
+                Some(info.clone())
+            })
+            .collect()
+    }
+
+    pub fn get_near_location(
+        &mut self,
+        loc: u32,
+        limit: usize,
+    ) -> Vec<Arc<AgentInfoSigned>> {
+        self.check_prune();
+
+        let mut out: Vec<(u32, &Arc<AgentInfoSigned>)> = self
+            .store
+            .values()
+            .filter_map(|v| {
+                if !v.is_tombstone {
+                    if v.storage_arc.is_none() {
+                        // filter out zero arcs, they can't help us
+                        None
+                    } else {
+                        Some((calc_dist(loc, v.storage_arc), v))
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        out.sort_by(|a, b| a.0.cmp(&b.0));
+
+        out.into_iter()
+            .take(limit)
+            .map(|(_, v)| v.clone())
+            .collect()
+    }
+}
+
+/// Get the min distance from a location to an arc in a wrapping u32 space.
+/// This function will only return 0 if the location is covered by the arc.
+/// This function will return u32::MAX if the arc is not set.
+fn calc_dist(loc: u32, arc: BasicArc) -> u32 {
+    match arc {
+        None => u32::MAX,
+        Some((arc_start, arc_end)) => {
+            let (d1, d2) = if arc_start > arc_end {
+                // this arc wraps around the end of u32::MAX
+
+                if loc >= arc_start || loc <= arc_end {
+                    return 0;
+                } else {
+                    (loc - arc_end, arc_start - loc)
+                }
+            } else {
+                // this arc does not wrap
+
+                if loc >= arc_start && loc <= arc_end {
+                    return 0;
+                } else if loc < arc_start {
+                    (arc_start - loc, u32::MAX - arc_end + loc + 1)
+                } else {
+                    (loc - arc_end, u32::MAX - loc + arc_start + 1)
+                }
+            };
+            std::cmp::min(d1, d2)
+        }
+    }
+}
+
+/// Determine if any part of two arcs overlap.
+fn arcs_overlap(a: BasicArc, b: BasicArc) -> bool {
+    match (a, b) {
+        (None, _) | (_, None) => false,
+        (Some((a_beg, a_end)), Some((b_beg, b_end))) => {
+            // The only way for there to be overlap is if
+            // either of a's start or end points are within b
+            // or either of b's start or end points are within a
+            calc_dist(a_beg, Some((b_beg, b_end))) == 0
+                || calc_dist(a_end, Some((b_beg, b_end))) == 0
+                || calc_dist(b_beg, Some((a_beg, a_end))) == 0
+                || calc_dist(b_end, Some((a_beg, a_end))) == 0
```

Comment:
Don't we only need the first 2 comparisons and the next 2 are redundant? Or am I missing a case?

Reply:
We need to check both sides, because they can fully contain each other, in which case one direction might not catch it:

```
|-A-B--B-A-|
vs
|-B-A--A-B-|
```

Reply:
You could add this to the function doc.

Reply:
Done: https://github.com/holochain/kitsune2/pull/19/commits/051985b666b17dfb07d172005ed233b021111cfe